krm-functions/lib/nad/v1/nad.go

```diff
@@ -0,0 +1,283 @@
+	nadConfigStruct := NadConfig{}
+	if err := json.Unmarshal([]byte(r.GetStringValue(ConfigType...)), &nadConfigStruct); err != nil {
+		return []Addresses{}
+	}
+	return nadConfigStruct.Plugins[0].Ipam.Addresses
+}
+
+func (r *NadStruct) GetNadConfig() NadConfig {
+	nadConfigStruct := NadConfig{}
+	configSpec := r.GetStringValue(ConfigType...)
+	if configSpec == "" {
+		configSpec = "{}"
+	}
+	if err := json.Unmarshal([]byte(configSpec), &nadConfigStruct); err != nil {
+		panic(err)
+	}
+	nadConfigStruct.CniVersion = CniVersion
+	if r.GetCniSpecType() == VlanType {
+		return nadConfigStruct
+	} else if r.GetCniSpecType() == IpamType || r.GetCniSpecType() == Both {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{
+						Ips: true,
+					},
+					Mode: NadMode,
+					Ipam: Ipam{
+						Type: NadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+			}
+		}
+		return nadConfigStruct
+	}
+	return nadConfigStruct
+}
+
+func (r *NadStruct) SetCniSpecType(cniType CniSpecType) {
+	r.C = cniType
+}
+
+// SetConfigSpec sets the spec attributes in the kubeobject according the go struct
+func (r *NadStruct) SetConfigSpec(spec *nadv1.NetworkAttachmentDefinitionSpec) error {
+	b, err := json.Marshal(spec.Config)
+	if err != nil {
+		return err
+	}
+	return r.K.SetNestedString(string(b), ConfigType...)
+}
+
+func (r *NadStruct) SetCNIType(cniType string) error {
+	if cniType != "" {
+		nadConfigStruct := r.GetNadConfig()
+		nadConfigStruct.Plugins[0].Type = cniType
+		b, err := json.Marshal(nadConfigStruct)
+		if err != nil {
+			return err
+		}
+		return r.K.SetNestedString(string(b), ConfigType...)
+	} else {
+		return fmt.Errorf("unknown cniType")
+	}
+}
+
+func (r *NadStruct) SetVlan(vlanType int) error {
+	if vlanType != 0 {
+		nadConfigStruct := r.GetNadConfig()
+		nadConfigStruct.Vlan = vlanType
+		b, err := json.Marshal(nadConfigStruct)
+		if err != nil {
+			return err
+		}
+		return r.K.SetNestedString(string(b), ConfigType...)
+	} else {
+		return fmt.Errorf("unknown vlanType")
+	}
+}
+
+func (r *NadStruct) SetNadMaster(nadMaster string) error {
+	if nadMaster != "" {
+		nadConfigStruct := r.GetNadConfig()
+		nadConfigStruct.Plugins[0].Master = nadMaster
+		b, err := json.Marshal(nadConfigStruct)
+		if err != nil {
+			return err
+		}
+		return r.K.SetNestedString(string(b), ConfigType...)
+	} else {
+		return fmt.Errorf("unknown nad master interface")
+	}
+}
+
+func (r *NadStruct) SetIpamAddress(ipam []Addresses) error {
+	if ipam != nil {
+		nadConfigStruct := r.GetNadConfig()
+		nadConfigStruct.Plugins[0].Ipam.Addresses = ipam
+		b, err := json.Marshal(nadConfigStruct)
+		if err != nil {
+			return err
+		}
+		return r.K.SetNestedString(string(b), ConfigType...)
+	} else {
+		return fmt.Errorf("unknown IPAM address")
+	}
+}
+
+func (n *NadConfig) ToString() string {
```

Comment:
Consider renaming this to String() to comply with Stringer interface: https://go.dev/tour/methods/17